### PR TITLE
828 save waveform constants with acquisition and generate method to load it

### LINKED ISSUE #239: Implement and test bidirectional ASLM
Triangle based scan instead of sawtooth.

Should also get a solid number on the bandwidth of our scan capabilities for both waveforms.

### LINKED ISSUE #828: Save Waveform Constants with Acquisition and Generate Method to Load it
We save the `experiment.yaml` file with every acquisition, allowing users to reload settings from previous experiments.

We should do the same with the `waveform_constants.yaml` file. Upon saving data, save the file. Also need a way to load it into the software as a menu item (just like the experiment files).

Menu item - File Load Waveform Constants -> `menu_controller`

Controller - `populate_experiment_setting`

Saving the waveform constants - `Image_writer` I think

### PR DESCRIPTION
**Pull request by AdvancedImagingUTSW**
Annie, this could use a close look. Sriya led the charge on these code changes, and her branch combined two different issues: #828 and #804 

For #828, she had a hard time getting the waveform popup to repopulate with new values, once she loaded a new waveform constants file. I dabbled with it, and then resorted to something that may not be ideal; I now destroy the waveform popup window, delete the waveform popup controller, and then launch a new popup and controller. To make it less annoying, The new popup shows up where the old one was. 

For #804, we will need to change things to accommodate all of the work you did overhauling the camera timing.

### CODE REVIEW

File: src/navigate/controller/sub_controllers/menu_controller.py [hunk 1]
```diff
@@ -108,27 +116,40 @@ def __init__(self, view, parent_controller=None):
...
+
         #: int: Feature list count.
         self.feature_list_count = 0
+
         #: str: Feature list file name.
         self.feature_list_file_name = "feature_lists.yaml"
 
+        #: waveform_popup_controller: Waveform Popup Controller.
+        # self.parent_controller.waveform_popup_controller = None
+
```
Comment (AdvancedImagingUTSW) [inline src/navigate/controller/sub_controllers/menu_controller.py]: Perhaps we should delete these commented lines.

File: src/navigate/model/devices/APIs/hamamatsu/HamamatsuAPI.py [hunk 2]
```diff
@@ -649,8 +650,8 @@ class DCAMDEV_STRING(Structure):
     "trigger_polarity": 1049120,  # 0x00100220, R/W, mode, "TRIGGER POLARITY"
     "trigger_source": 1048848,  # 0x00100110, R/W, mode,   "TRIGGER SOURCE"
     "trigger_delay": 1049184,  # 0x00100260,	/* R/W, sec,	"TRIGGER DELAY"
-    "internal_line_interval": 4208720,  # 0x00403850, R/W, sec,
-    # "INTERNAL LINE INTERVAL"
+    "internal_line_interval": 4208720,  # 0x00403850, R/W, sec,    "INTERNAL LINE INTERVAL"
+    "internal_line_speed": 4208704,  # 0x00403840, R/W, m/sec,    "INTERNAL LINE SPEED"
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/APIs/hamamatsu/HamamatsuAPI.py]: Important changes, presumably.

File: src/navigate/model/devices/camera/camera_base.py [hunk 3]
```diff
@@ -83,10 +83,6 @@ def __init__(self, microscope_name, device_connection, configuration):
         self.is_acquiring = False
 
         # Initialize Pixel Information
-        #: int: Maximum image width
-        self.max_image_width = 2048
-        #: int: Maximum image height
-        self.max_image_height = 2048
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/camera/camera_base.py]: More important stuff.

File: src/navigate/model/devices/camera/camera_base.py [hunk 4]
```diff
@@ -96,11 +92,13 @@ def __init__(self, microscope_name, device_connection, configuration):
...
         self.step_image_height = 4
         #: int: Number of pixels in the x direction
-        self.x_pixels = self.max_image_width
+        self.x_pixels = 2048
         #: int: Number of pixels in the y direction
-        self.y_pixels = self.max_image_height
-        self.camera_parameters["x_pixels"] = self.max_image_width
-        self.camera_parameters["y_pixels"] = self.max_image_height
+        self.y_pixels = 2048
+        self.camera_parameters["x_pixels"] = 2048
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/camera/camera_base.py]: Hard coding

File: src/navigate/model/devices/camera/camera_hamamatsu.py [hunk 5]
```diff
@@ -43,10 +43,11 @@
 logger = logging.getLogger(p)
 
 
-class HamamatsuOrca(CameraBase):
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/camera/camera_hamamatsu.py]: I assume that most of these changes can be ignored. I take responsibility for this, and I'm sorry.

File: src/navigate/model/devices/daq/daq_ni.py [hunk 6]
```diff
@@ -264,6 +266,9 @@ def create_analog_output_tasks(self, channel_key):
         """
         self.n_sample = int(self.sample_rate * self.sweep_times[channel_key])
         max_sample = self.n_sample * self.waveform_expand_num
+        print("max_sample", max_sample)
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/daq/daq_ni.py]: Delete print statements.

File: src/navigate/model/devices/galvo/galvo_base.py [hunk 7]
```diff
@@ -188,7 +188,7 @@ def adjust(self, exposure_times, sweep_times):
                 elif self.galvo_waveform == "halfsaw":
                     new_wave = sawtooth(
                         sample_rate=self.sample_rate,
-                        sweep_time=self.sweep_time,
+                        sweep_time= self.sweep_time,
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/devices/galvo/galvo_base.py]: Uniform spacing

File: src/navigate/model/microscope.py [hunk 8]
```diff
@@ -405,13 +405,18 @@ def prepare_acquisition(self):
         sensor_mode = self.configuration["experiment"]["CameraParameters"][
             "sensor_mode"
         ]
+
+        print('sensor_mode', sensor_mode)
+
```
Comment (AdvancedImagingUTSW) [inline src/navigate/model/microscope.py]: Delete print statement

**Review by annie-xd-wang — APPROVED** [review]
We can test bidirectional acquisition on a machine now.

File: src/navigate/model/microscope.py (hunk 8 above)
Comment (annie-xd-wang) [reply src/navigate/model/microscope.py]: Deleted.

File: src/navigate/model/devices/galvo/galvo_base.py (hunk 7 above)
Comment (annie-xd-wang) [reply src/navigate/model/devices/galvo/galvo_base.py]: Fixed.

File: src/navigate/model/devices/daq/daq_ni.py (hunk 6 above)
Comment (annie-xd-wang) [reply src/navigate/model/devices/daq/daq_ni.py]: Deleted.

File: src/navigate/controller/sub_controllers/menu_controller.py (hunk 1 above)
Comment (annie-xd-wang) [reply src/navigate/controller/sub_controllers/menu_controller.py]: Deleted.

File: src/navigate/model/devices/APIs/hamamatsu/HamamatsuAPI.py (hunk 2 above)
Comment (annie-xd-wang) [reply src/navigate/model/devices/APIs/hamamatsu/HamamatsuAPI.py]: The modifications to camera functions originate from the commits in PR #808.

File: src/navigate/controller/sub_controllers/menu_controller.py [hunk 9]
```diff
@@ -690,6 +722,45 @@ def save_experiment(self, *args):
...
+            self.parent_controller.configuration,
+            "waveform_constants",
+            filename,
+        )
+        verify_waveform_constants(
+            self.parent_controller.manager, self.parent_controller.configuration
+        )
+
+        if hasattr(self.parent_controller, "waveform_popup_controller"):
+            self.parent_controller.waveform_popup_controller.populate_experiment_values(
```
Comment (AdvancedImagingUTSW) [inline src/navigate/controller/sub_controllers/menu_controller.py:760]: Much nicer.

**Comment by AdvancedImagingUTSW** [comment]
Will test first thing tomorrow. But everything looks great.

**Pull request merged**